src/cmap/establish/handshake/mod.rs

```diff
@@ -0,0 +1,192 @@
+#[cfg(test)]
+mod test;
+
+use bson::{doc, Bson, Document};
+use lazy_static::lazy_static;
+use os_info::{Type, Version};
+
+use crate::{
+    cmap::{options::ConnectionPoolOptions, Command, Connection, StreamDescription},
+    error::Result,
+    is_master::IsMasterReply,
+};
+
+#[derive(Clone, Debug)]
+struct ClientMetadata {
+    application: Option<AppMetadata>,
+    driver: DriverMetadata,
+    os: OsMetadata,
+    platform: Option<String>,
+}
+
+#[derive(Clone, Debug)]
+struct AppMetadata {
+    name: String,
+}
+
+#[derive(Clone, Debug)]
+struct DriverMetadata {
+    name: String,
+    version: String,
+}
+
+#[derive(Clone, Debug)]
+struct OsMetadata {
+    os_type: String,
+    name: Option<String>,
+    architecture: String,
+    version: Option<String>,
+}
+
+impl From<ClientMetadata> for Bson {
+    fn from(metadata: ClientMetadata) -> Self {
+        let mut metadata_doc = Document::new();
```

Comment:
While we'd normally use `Serialize` and just propagate the errors from `bson::to_bson` for this, the place where `Handshaker::new` is actually called is often in an SDAM background task (as the `Establisher` gets created when a pool is created, which in turn is created when a new server is created, which only happens either at client creation or when a new server is discovered). Since we don't really have any way to handle a failure to create a connection pool there, I think it's cleaner to just manually write the infallible code for converting this to a document rather than either calling `unwrap` or encoding error-handling logic in the SDAM tasks where we errors won't actually ever occur.

In the longer term, it's worth looking into if there's any way for us to implement infallible conversion in the BSON library for types that are guaranteed to convert safely.